## Welcome

I work for [HPCBio](https://hpcbio.illinois.edu/) at the University of Illinois,
Urbana-Champaign, where I
perform bioinformatics analysis, consultation, and training.  My background is
in plant population genetics and breeding. I am a developer of software
for the genetic analysis of polyploid organisms.  I earned a PhD in Genetics
from UC Davis in 2011 and a BA in Genetics, Cell & Developmental Biology from
Dartmouth College in 2004.

[LinkedIn](https://www.linkedin.com/in/lindsay-clark-75539aa9/)

[ORCID](https://orcid.org/0000-0002-3881-9252)

[Google Scholar](https://scholar.google.com/citations?user=d2OkHU4AAAAJ)

[GitHub](https://github.com/lvclark/)

[Publons](https://publons.com/researcher/734253/lindsay-virginia-clark/)

## Excellence in Breeding meeting May 8-10, 2019

Datasets for tutorials are available at
<https://gobiiproject.atlassian.net/wiki/spaces/WG/pages/70451254/Data+sets>

Wednesday, May 8

* Tools for variant and genotype calling -- [Powerpoint](excellence_in_breeding/Variant_and_Genotype_Calling_in_Polyploids.pdf)
* File formats for polyploid genotypes -- [Powerpoint](excellence_in_breeding/File_formats_for_polyploid_genotypes.pdf)
* Genotype calling with polyRAD in a sweetpotato F1 population -- [Tutorial on web](excellence_in_breeding/sweetpotato_F1_polyRAD_tutorial.html) -- [Download R markdown](excellence_in_breeding/sweetpotato_F1_polyRAD_tutorial.Rmd)
* Genotype calling with polyRAD in a *Miscanthus* diversity panel -- [Tutorial on web](excellence_in_breeding/polyRAD_diversity_panel.html) -- [Download R markdown](excellence_in_breeding/polyRAD_diversity_panel.Rmd)

Thursday, May 9

* GWAS in GAPIT with posterior mean genotypes -- [Tutorial on web](excellence_in_breeding/GAPIT_tetraploid.html) -- [Download R markdown](excellence_in_breeding/GAPIT_tetraploid.Rmd)
* Genomic prediction in rrBLUP using posterior mean genotypes -- [Tutorial on web](excellence_in_breeding/rrBLUP_tetraploid.html) -- [Download R markdown](excellence_in_breeding/rrBLUP_tetraploid.Rmd)

Friday, May 10

* [Software Carpentry lesson](https://swc-uiuc.github.io/2019-05-10-cip/)
* Calling variants without a reference genome using GBS-SNP-CROP -- [Tutorial on web](excellence_in_breeding/calling_without_reference.html) -- [Download markdown](https://raw.githubusercontent.com/lvclark/lvclark.github.io/master/excellence_in_breeding/calling_without_reference.md)
* Exploring and filtering VCFs using Bioconductor -- [Tutorial on web](excellence_in_breeding/vcf_filter.html) -- [Download R Markdown](excellence_in_breeding/vcf_filter.Rmd)
